base/ltcmdhooks.dtx

```diff
@@ -577,20 +577,53 @@
 %   otherwise it tries to update the hook.
 % \changes{v1.0d}{2021/08/25}{Simplify generic hook detection}
 %    \begin{macrocode}
-%<latexrelease>\IncludeInRelease{2021/11/15}{\@@_try_put_cmd_hook:n}%
-%<latexrelease>                 {Standardise~generic~hook~names}
+%<latexrelease>\IncludeInRelease{2024/12/22}{\@@_try_put_cmd_hook:n}%
+%<latexrelease>                 {Don't~define~command}
 \cs_new_protected:Npn \@@_try_put_cmd_hook:n #1
   { \@@_try_put_cmd_hook:w #1 / / / \s_@@_mark {#1} }
 \cs_new_protected:Npn \@@_try_put_cmd_hook:w
     #1 / #2 / #3 / #4 \s_@@_mark #5
   {
     \@@_debug:n { \iow_term:n { ->~Adding~cmd~hook~to~'#2'~(#3): } }
-    \exp_args:Nc \@@_patch_cmd_or_delay:Nnn {#2} {#2} {#3}
+%    \end{macrocode}
+%    \cs{@@_patch_cmd_or_delay:Nnn} expects the command to be patched
+%    as its first argument so we need to construct it from its name
+%    (\texttt{\#2}). However, at this moment it may not exist yet, so
+%    using \cs{cs:w} would incorrectly turn it from
+%    \enquote{undefined} into \cs{relax}. We  therefore use the
+%    following curious construction: we start a group and expand out of
+%    it to call \cs{cs:w}. If the command is now changed to \cs{relax}
+%    the \cs{group_end:} will undo that change, but the token is
+%    nevertheless there to be consumed by \cs{@@_patch_cmd_or_delay:Nnn}.
+% \changes{v1.0k}{2024/12/22}{Avoid defining command while adding a cmd
+%    hook (gh/1591)}
+%    \begin{macrocode}
+    \group_begin:
+       \exp_after:wN
+    \group_end:
+       \exp_after:wN
+    \@@_patch_cmd_or_delay:Nnn
+       \cs:w #2\cs_end:
```

Comment:
Maybe use `\group_begin: \exp_args:NNc \group_end:` instead of the `\exp_after:wN` chain.

Reply:
I'm not sure it is more readable this way (given that it is a hack anyway)